hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AFileSystem.java

```diff
@@ -891,50 +902,123 @@ private boolean innerRename(Path source, Path dest)
+                            new DeleteObjectsRequest.KeyVersion(key), deleteQueue, copyFailure)),
+                    key, newDstKey, length));
           } else {
-            S3Guard.addMoveFile(metadataStore, srcPaths, dstMetas, childSrc,
-                childDst, length, getDefaultBlockSize(childDst), username);
+            // We got a copy failure, so don't bother going through the rest of the files
+            break;
           }
-          // Ancestor directories may not be listed, so we explicitly add them
-          S3Guard.addMoveAncestors(metadataStore, srcPaths, dstMetas,
-              keyToQualifiedPath(srcKey), childSrc, childDst, username);
         }
 
-        if (keysToDelete.size() == MAX_ENTRIES_TO_DELETE) {
-          removeKeys(keysToDelete, true, false);
+        for (CopyContext copyContext : copies) {
+          try {
+            copyContext.getCopy().waitForCopyResult();
+          } catch (InterruptedException e) {
+            throw new RenameFailedException(copyContext.getSrcKey(), copyContext.getDstKey(), e);
+          }
+
+          if (hasMetadataStore()) {
+            // with a metadata store, the object entries need to be updated,
+            // including, potentially, the ancestors
+            Path childSrc = keyToQualifiedPath(copyContext.getSrcKey());
+            Path childDst = keyToQualifiedPath(copyContext.getDstKey());
+            if (objectRepresentsDirectory(copyContext.getSrcKey(), copyContext.getLength())) {
+              S3Guard.addMoveDir(metadataStore, srcPaths, dstMetas, childSrc,
+                      childDst, username);
+            } else {
+              S3Guard.addMoveFile(metadataStore, srcPaths, dstMetas, childSrc,
+                      childDst, copyContext.getLength(), getDefaultBlockSize(childDst), username);
+            }
+            // Ancestor directories may not be listed, so we explicitly add them
+            S3Guard.addMoveAncestors(metadataStore, srcPaths, dstMetas,
+                    keyToQualifiedPath(srcKey), childSrc, childDst, username);
+          }
+        }
+
+        if (copyFailure.get()) {
+          throw new RenameFailedException(srcKey, dstKey,
+                  new IllegalStateException("Progress listener indicated a copy failure, but no exception was thrown"));
+        }
+
+        try {
+          for (DeleteObjectsRequest.KeyVersion dirKey : dirKeysToDelete) {
+            deleteQueue.put(Optional.of(dirKey));
+          }
+          deleteQueue.put(Optional.empty());
+          deleteFuture.get();
+        } catch (ExecutionException | InterruptedException e) {
+          throw new RenameFailedException(srcKey, dstKey, e);
+        }
+        deleteFutureComplete = true;
+      } finally {
+        if (!deleteFutureComplete) {
+          if (deleteFuture != null && !deleteFuture.isDone() && !deleteFuture.isCancelled()) {
+            deleteFuture.cancel(true);
+          }
```

Comment:
Wondering if we can move parallel rename logic to a new class.. S3AFileSystem is getting too big.

Reply:
mmm. I mostly concur, handing in either the S3aFS or the (expanded) WriteOperationsHelper

Reply:
Moved all the parallel rename logic into a dedicated calls called `ParallelDirectoryRenamer`